### DISCUSSION TITLE
Models with Optional fields raise TypeError: issubclass() arg 1 must be a class

### DISCUSSION THREAD

**Discussion by alexiob** [fastapi/sqlmodel#777] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the SQLModel documentation, with the integrated search.
- [X] I already searched in Google "How to X in SQLModel" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to SQLModel but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to SQLModel but to [SQLAlchemy](https://github.com/sqlalchemy/sqlalchemy).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from pydantic import UUID4
from sqlmodel import Field
from sqlmodel import SQLModel


class UserRoleLink(SQLModel, table=True):  # type: ignore
    __tablename__ = "user_role_links"

    user_id: UUID4 | None = Field(default=None, foreign_key="user.id", primary_key=True)
    role_id: UUID4 | None = Field(
        default=None, foreign_key="roles.id", primary_key=True
    )
```

### Description

When a model has one or more Optional fields, SQLModel raises an error. If the Optionals are removed, it works fine. It happens for all the defined models using Optionals. Also if the code is changed to use the `Optional[]` format.

It errors when starting the application and when running `pytest`.

The traceback:

```
<sqlmodel._compat.FakeMetadata object at 0x109743650> annotation=Union[Annotated[UUID, UuidVersion], NoneType] required=False
ImportError while loading conftest '/backend/conftest.py'.
conftest.py:12: in <module>
    from quantum.auth.roles import RoleManager
quantum/auth/roles.py:19: in <module>
    from quantum.auth.scopes import Scope
quantum/auth/scopes.py:8: in <module>
    from quantum.data import schemas
quantum/data/schemas/__init__.py:4: in <module>
    from .analytics import AnalyticsAggregateDailyPdfGenerationsRead
quantum/data/schemas/analytics/__init__.py:1: in <module>
    from .daily_projects import AnalyticsDailyProjectsRead
quantum/data/schemas/analytics/daily_projects.py:5: in <module>
    from quantum.data.models.analytics.daily_projects import OrgProjects
quantum/data/models/__init__.py:13: in <module>
    from .role import Role
quantum/data/models/role.py:14: in <module>
    from .user_role_link import UserRoleLink
quantum/data/models/user_role_link.py:6: in <module>
    class UserRoleLink(SQLModel, table=True):  # type: ignore
/Users/alex/Library/Caches/pypoetry/virtualenvs/quantum-backend-Gxs7_MCe-py3.11/lib/python3.11/site-packages/sqlmodel/main.py:474: in __new__
    col = get_column_from_field(v)
/Users/alex/Library/Caches/pypoetry/virtualenvs/quantum-backend-Gxs7_MCe-py3.11/lib/python3.11/site-packages/sqlmodel/main.py:618: in get_column_from_field
    sa_type = get_sqlalchemy_type(field)
/Users/alex/Library/Caches/pypoetry/virtualenvs/quantum-backend-Gxs7_MCe-py3.11/lib/python3.11/site-packages/sqlmodel/main.py:567: in get_sqlalchemy_type
    if issubclass(type_, Enum):
E   TypeError: issubclass() arg 1 must be a class
```

I have added a simple print in the `def get_sqlalchemy_type(field: Any) -> Any:` function, just before the `issubclass` checks:

```
>>> INFO typing.Annotated[uuid.UUID, UuidVersion(uuid_version=4)] <class 'typing._AnnotatedAlias'> 
```

 It happens with SQLModel `0.0.14` and with the current `master`.

### Operating System

macOS

### SQLModel Version

0.0.14

### Python Version

Python 3.11.0

### Additional Context

Looks similar to https://github.com/tiangolo/sqlmodel/issues/57

**Comment by hakkano** [comment]
I am facing a similar issue with Enums 


```
class XEnum (str, Enum):
  a = "a"
  b = "b"

class XTable(SQLModel, table = true):
    fielda: str = Field(
        default=None,
        primary_key=True,
    )
    fieldb: XEnum = Field(primary_key=True)
    fieldc: Optional[str] = Field(default = None)
```
where I get error like this

```
 File "/redacted/utilities/app/models/gstModels.py", line 296, in <module>
    class XTable(SQLModel, table=True):
  File "/redacted/utilities/env/lib/python3.10/site-packages/sqlmodel/main.py", line 474, in __new__
    col = get_column_from_field(v)
  File "/redacted/utilities/env/lib/python3.10/site-packages/sqlmodel/main.py", line 618, in get_column_from_field
    sa_type = get_sqlalchemy_type(field)
  File "/redacted/utilities/env/lib/python3.10/site-packages/sqlmodel/main.py", line 567, in get_sqlalchemy_type
    if issubclass(type_, Enum):
TypeError: issubclass() arg 1 must be a class
```

**Reply by hakkano** [reply]
Also, checking if this works https://fastapi.tiangolo.com/python-types/#using-union-or-optional

**Reply by alexiob** [reply]
I have tried with `Union` as well: same error

**Comment by ChrisNi888** [comment]
I have the same error. (using Pydantic 2.7 and SQLModel 0.0.16. ) So I use Pydantic 1.0.11 and SQLModel 0.08. It works. If anyone has same problem, hope it helps.

**Comment by joachimhuet** [comment]
Hello there, in fact this error is triggered by sqlalchemy. If you remove the `table=True` it should work.
One way I found to get it working is to define the `sa_type` for the concerned column:

@alexiob 
```python
from sqlalchemy import Uuid

class UserRoleLink(SQLModel, table=True):
    __tablename__ = "user_role_links"

    user_id: Optional[UUID4] = Field(default=None, foreign_key="user.id", primary_key=True, sa_type=Uuid)
    role_id: Optional[UUID4] = Field(
        default=None, foreign_key="roles.id", primary_key=True, sa_type=Uuid
    )
```

@hakkano 
Your exemple works for me, maybe install latest version.

@ChrisNi888 
You should get back to pydantic v2 🤔

**Reply by ChrisNi888** [reply]
Yes. I found that if I define the sa_type for HttpUrl attribute, it will be fine.
`avatar: Optional[HttpUrl] = Field(sa_type=AutoString)`
Thanks!

**Comment by DamonGuzman** [comment]
For me, the issue was caused by using `typing.Literal`

**Model:**
``` python
class AccessControlDB(SQLModel, table=True):
    __tablename__ = "access_control"

    id: UUID = Field(default=uuid4, primary_key=True)
    created_at: datetime | None= Field(default=None)
    attribute: str
    operation: typing.Literal['CREATE', 'UPDATE', 'DELETE', 'READ']
    user_id: UUID
```

The error is specifically caused by this line:
`operation: typing.Literal['CREATE', 'UPDATE', 'DELETE', 'READ']`

**Error:**
``` shell
  File "<redacted_path>\database\models\attribute.py", line 17, in <module>
    class AccessControlDB(SQLModel, table=True):
  File "<redacted_path>\site-packages\sqlmodel\main.py", line 482, in __new__
    col = get_column_from_field(v)
          ^^^^^^^^^^^^^^^^^^^^^^^^
  File "<redacted_path>\site-packages\sqlmodel\main.py", line 626, in get_column_from_field
    sa_type = get_sqlalchemy_type(field)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<redacted_path>\site-packages\sqlmodel\main.py", line 575, in get_sqlalchemy_type
    if issubclass(type_, Enum):
       ^^^^^^^^^^^^^^^^^^^^^^^
TypeError: issubclass() arg 1 must be a class
```

My current interpretation is that SQLModel is implicitly converting a typing.Literal into an Enum but it's not being handled correctly. 


I resolved this error by changing my implementation to using an actual enum.Enum

---

**Solution:**
``` python
class OperationEnum(Enum):
    CREATE = 'CREATE'
    UPDATE = 'UPDATE'
    DELETE = 'DELETE'
    READ = 'READ'

class AccessControlDB(SQLModel, table=True):
    __tablename__ = "access_control"

    id: UUID = Field(default=uuid4, primary_key=True)
    created_at: datetime | None= Field(default=None)
    attribute: str
    operation: OperationEnum
    user_id: UUID
```

**Reply by joachimhuet** [reply]
Well your problem is indeed the same, sqlalchemy doesn't know to what type your Literal should be parsed to in the database. Instead of changing it to an enum, you could also specify `sa_type=String` and it should work. 
It's just about giving sqlalchemy the information of the type it should use.

**Reply by DamonGuzman** [reply]
> use.

Why would I use a string? I was originally using a typing.Literal and the closest thing to that is an Enum. 

I'm trying to represent one of a few predetermined options.

**Reply by joachimhuet** [reply]
> My current interpretation is that SQLModel is implicitly converting a typing.Literal into an Enum but it's not being handled correctly.

What I am trying to say is that your statement above is not correct, and that is why the error is fired. If it converted the Literal type to an enum it would work fine.
And the important thing to understand here is that some types that are not natively recognized by sqlalchemy will need to have a type specification. And that is exactly the problem you faced with that, there is no straight forward equivalent in the database model for a `typing.Literal` python type.
So specifying `sa_type=String` will create a database column typed as a string whereas changing the type to enum will use the Enum type if it exists or a string (nvarchar) if not (and by default won't add a check constraint). So for your case you are probably right to choose the enum. But changing the type is not a solution to the problem but just avoiding it, if you want to keep the `Literal` type you can make it a string and add a `CheckConstraint`. But of course going with an enum makes a lot more sense.
Hope this small explanation helped you understand the difference between the class object type and the sqlalchemy type that will then be used to create your table column.

**Comment by vdozier-work** [comment]
Same here, I just posted this https://github.com/tiangolo/sqlmodel/discussions/981

I didn't see this thread when I was first searching

**Comment by YuriiMotov** [comment]
`UUID4 | None` does not issue an error in the current version (0.0.24). It looks like it's fixed.
